internal/plugin/metrics.go

```diff
@@ -0,0 +1,303 @@
+		// MessagePipe using the metrics topic, as they will be handled in the metrics plugin.
+		Start(ctx context.Context, updateInterval time.Duration) error
+		// Stops the data source from collecting data.
+		Stop()
+		// Getter for the type of data source.
+		Type() string
+	}
+
+	// Tuple for a data source and its associated OTel Meter.
+	MetricsSource struct {
+		dataSource DataSource
+		meter      metric.Meter
+	}
+
+	// The Metrics plugin. Discovers and owns the data sources that produce metrics for the Agent.
+	Metrics struct {
+		Sources       map[string]*MetricsSource    // key = DataSource.Type
+		counterCache  map[string]metrics.DataPoint // key = DataEntry.Name
+		conf          config.Config
+		pipe          bus.MessagePipeInterface
+		prod          *metrics.MetricsProducer
+		shutdownFuncs []shutdownFunc
+	}
+
+	// The Metrics plugin can be configured with functional options.
+	MetricsOption func(*Metrics) error
+)
+
+// Constructor for the Metrics plugin.
+func NewMetrics(c config.Config, options ...MetricsOption) (*Metrics, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	prod := metrics.NewMetricsProducer(c.Version)
+	meterProvider, err := metrics.NewMeterProvider(ctx, otelServiceName, *c.Metrics, prod)
+	if err != nil {
+		log.Printf("failed to create a meterProvider: %v", err)
+	}
+
+	// Sets the global meter provider.
+	otel.SetMeterProvider(meterProvider)
+
+	m := Metrics{
+		Sources:       make(map[string]*MetricsSource, 0), // Key = data source type
+		counterCache:  make(map[string]metrics.DataPoint, 0),
+		conf:          c,
+		pipe:          nil,
+		prod:          prod,
+		shutdownFuncs: []shutdownFunc{shutdownFunc(cancel)},
+	}
+
+	for _, opt := range options {
+		err := opt(&m)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply metrics plugin option: %w", err)
+		}
+	}
+
+	return &m, nil
+}
+
+// Initializes and starts the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Init(mp bus.MessagePipeInterface) error {
+	m.pipe = mp
+	go m.prod.StartListen(mp.Context())
+
+	err := m.discoverSources()
+	if err != nil {
+		return fmt.Errorf("data source discovery failed: %w", err)
+	}
+
+	return m.startDataSources()
+}
+
+// Stops the metrics plugin. Required for the `Plugin` interface.
+func (m *Metrics) Close() error {
+	slog.Info("stopping all metrics data sources")
+	for _, ms := range m.Sources {
+		ms.dataSource.Stop()
+	}
+
+	for _, shutdown := range m.shutdownFuncs {
+		shutdown()
+	}
+
+	return nil
+}
+
+// Returns info about the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Info() *bus.Info {
+	return &bus.Info{
+		Name: "metrics",
+	}
+}
+
+// Processes an incoming Message Bus message in the plugin. Required for the `Plugin` interface.
+func (m *Metrics) Process(msg *bus.Message) {
+	switch msg.Topic {
+	case bus.METRICS_TOPIC:
+		entry, ok := msg.Data.(metrics.DataEntry)
+		if !ok {
+			slog.Warn("failed to convert metrics event to a data entry")
```

Comment:
If the datatype is wrong we should return at this stage
```suggestion
			slog.Warn("Failed to convert metrics event to a data entry")
			return
```